module/x/gravity/keeper/attestation_handler.go

```diff
@@ -21,317 +22,518 @@ var _ types.DistributionKeeper = (*distrkeeper.Keeper)(nil)
+
+	if claim.Symbol != metadata.Symbol {
+		return sdkerrors.Wrap(
+			types.ErrInvalid,
+			fmt.Sprintf("ERC20 symbol %s does not match denom symbol %s", claim.Symbol, metadata.Display))
+	}
+
+	// ERC20 tokens use a very simple mechanism to tell you where to display the decimal point.
+	// The "decimals" field simply tells you how many decimal places there will be.
+	// Cosmos denoms have a system that is much more full featured, with enterprise-ready token denominations.
+	// There is a DenomUnits array that tells you what the name of each denomination of the
+	// token is.
+	// To correlate this with an ERC20 "decimals" field, we have to search through the DenomUnits array
+	// to find the DenomUnit which matches up to the main token "display" value. Then we take the
+	// "exponent" from this DenomUnit.
+	// If the correct DenomUnit is not found, it will default to 0. This will result in there being no decimal places
+	// in the token's ERC20 on Ethereum. So, for example, if this happened with Atom, 1 Atom would appear on Ethereum
+	// as 1 million Atoms, having 6 extra places before the decimal point.
+	// This will only happen with a Denom Metadata which is for all intents and purposes invalid, but I am not sure
+	// this is checked for at any other point.
+	decimals := uint32(0)
+	for _, denomUnit := range metadata.DenomUnits {
+		if denomUnit.Denom == metadata.Display {
+			decimals = denomUnit.Exponent
+			break
 		}
+	}
 
-		if !invalidAddress { // valid address so far, try to lock up the coins in the requested cosmos address
-			if err := a.bankKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, nativeReceiver, coins); err != nil {
-				// someone attempted to send tokens to a blacklisted user from Ethereum, log and send to Community pool
-				hash, _ := claim.ClaimHash()
-				a.keeper.logger(ctx).Error("Blacklisted deposit",
-					"cause", err.Error(),
-					"claim type", claim.GetType(),
-					"id", types.GetAttestationKey(claim.GetEventNonce(), hash),
-					"nonce", fmt.Sprint(claim.GetEventNonce()),
+	if decimals != uint32(claim.Decimals) {
+		return sdkerrors.Wrap(
+			types.ErrInvalid,
+			fmt.Sprintf("ERC20 decimals %d does not match denom decimals %d", claim.Decimals, decimals))
+	}
+
+	// Add to denom-erc20 mapping
+	a.keeper.setCosmosOriginatedDenomToERC20(ctx, claim.CosmosDenom, *tokenAddress)
+
+	ctx.EventManager().EmitTypedEvent(
+		&types.EventERC20DeployedClaim{
+			Token: tokenAddress.GetAddress().Hex(),
+			Nonce: strconv.Itoa(int(claim.GetEventNonce())),
+		},
+	)
+	return nil
+}
+
+// Upon acceptance of sufficient ValsetUpdated claims: update LastObservedValset, mint cosmos-originated relayer rewards
+// so that the reward holder can send them to cosmos
+func (a AttestationHandler) handleValsetUpdated(ctx sdk.Context, claim types.MsgValsetUpdatedClaim) error {
+	rewardAddress, err := types.NewEthAddress(claim.RewardToken)
+	if err != nil {
+		return sdkerrors.Wrap(err, "invalid reward token on claim")
+	}
+
```

Comment:
Per our discussion @jkilpatr, I added the old valset's height. Very grateful for the exhaustive struct linter for helping us discover this issue. I also handled the TODO comment, I opted to panic but I am open to alternatives

Reply:
panic is good, lets us figure out what to do after.